engine/src/main/scala/cromwell/engine/workflow/lifecycle/WorkflowLifecycleActor.scala

```diff
@@ -0,0 +1,111 @@
+package cromwell.engine.workflow.lifecycle
+
+import akka.actor.{FSM, ActorRef, LoggingFSM}
+import cromwell.engine.workflow.lifecycle.WorkflowLifecycleActor._
+import wdl4s.Call
+
+object WorkflowLifecycleActor {
+
+  trait WorkflowLifecycleActorState { def terminal = false }
+  trait WorkflowLifecycleActorTerminalState extends WorkflowLifecycleActorState { override val terminal = true }
+
+  trait WorkflowLifecycleSuccessResponse
+  trait WorkflowLifecycleFailureResponse
+  trait WorkflowLifecycleAbortedResponse
+
+  /**
+    * State data
+    */
+  trait WorkflowLifecycleActorData{
+    /**
+      * The mapping from ActorRef to the backend-specific actor for that backend
+      */
+    def backendActors: Map[ActorRef, String]
+
+    /**
+      * The list of successful backend-specific actors
+      */
+    def successes: Seq[ActorRef]
+
+    /**
+      * The list of failed backend-specific actors
+      */
+    def failures: Map[ActorRef, Throwable]
+  }
+}
+
+trait WorkflowLifecycleActor[S <: WorkflowLifecycleActorState, D <: WorkflowLifecycleActorData] extends LoggingFSM[S, D] {
+
+  val successState: S
+  val failureState: S
+  val abortedState: S
+
+  def successResponse: WorkflowLifecycleSuccessResponse
+  def failureResponse(reasons: Seq[Throwable]): WorkflowLifecycleFailureResponse
+
+  when(successState) { FSM.NullFunction }
+  when(failureState) { FSM.NullFunction }
+  when(abortedState) { FSM.NullFunction }
+
+  whenUnhandled {
+    case unhandledMessage =>
+      log.warning(s"received an unhandled message: $unhandledMessage")
+      stay
+  }
+
+  onTransition {
+    case _ -> state if state.terminal =>
+      log.debug("State is now terminal. Shutting down.")
+      context.stop(self)
+    case fromState -> toState =>
+      // Only log this at debug - these states are never seen or used outside of the CallActor itself.
+      log.debug(s"State is transitioning from $fromState to $toState.")
+  }
+
+  /**
+    * Turns a mapping from Call => backendName into a mapping from backendName => List[Call]
+    */
+  protected def callAssignments(backendAssignments: Map[Call, String]): Map[String, List[Call]] = {
+    backendAssignments.foldLeft(Map.empty[String, List[Call]]) {
```

Comment:
Did you make that work? I got problems because backendAssignments wasn't a Seq and so didn't have `groupBy` - but if you got it to work I think that way's more intuitive